### PR TITLE
factplusplus: init at 1.6.5

### PR DESCRIPTION
**Pull request by mgttlinger**
###### Things done

Packaged the FaCT++ reasoner as a widespread OWL reasoner.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/science/logic/factplusplus/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{ stdenv, lib, fetchFromBitbucket, cmake, jdk }:
+
+stdenv.mkDerivation rec {
+
+  pname = "factplusplus";
+  version = "1.6.5";
+  name = "${pname}-${version}";
```
Comment (applePrincess) [inline pkgs/applications/science/logic/factplusplus/default.nix:7]: This line is unnecessary.

File: pkgs/applications/science/logic/factplusplus/default.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  installPhase = ''
+    mkdir -p $out/bin
+    mkdir -p $out/lib
+    install -m 755 FaCT++.{C,JNI,KE,Kernel}/obj/*.{so,o} $out/lib/
+    install -m 755 FaCT++/obj/FaCT++ $out/bin/FaCT++
+  '';
+
+   meta = with lib; {
+    description = "FaCT++ reasoner";
+    homepage    = http://owl.cs.manchester.ac.uk/tools/fact/;
```
Comment (applePrincess) [inline pkgs/applications/science/logic/factplusplus/default.nix:34]: URL should be quoted.

**Comment by mgttlinger** [comment]
Thanks for the review and the improvements @applePrincess!

File: pkgs/applications/science/logic/factplusplus/default.nix [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    mkdir -p $out/lib
+    install -m 755 FaCT++.{C,JNI,KE,Kernel}/obj/*.{so,o} $out/lib/
+    install -m 755 FaCT++/obj/FaCT++ $out/bin/FaCT++
+  '';
+
+   meta = with lib; {
+    description = "Tableaux-based reasoner for expressive Description Logics (DL)";
+    homepage    = "http://owl.cs.manchester.ac.uk/tools/fact/";
+    maintainers = [ maintainers.mgttlinger ];
+    license     = licenses.gpl2;
```
Comment (applePrincess) [inline pkgs/applications/science/logic/factplusplus/default.nix:34]: Sorry, I could have said better... `licenses.gpl2` is also deprecated (one of the reason is lack of clarity, eg whether later version is allowed or not). so better option might be `gpl2Only`, `gpl2Plus`, `lgpl2Only`, `lgpl2Plus`.
you can check its status by looking source code https://github.com/NixOS/nixpkgs/blob/master/lib/licenses.nix

Comment (mgttlinger) [reply pkgs/applications/science/logic/factplusplus/default.nix]: Thanks for clarifying. Fixed it in the latest commit.

**Comment by applePrincess** [comment]
if you squash into one

```
factplusplus: init at 1.6.5
```

looks good to me.

**Comment by applePrincess** [comment]
Result of `nixpkgs-review pr 194714` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>factplusplus</li>
  </ul>
</details>

**Review by applePrincess — APPROVED** [review]

**Comment by applePrincess** [comment]
```
[nix-shell:~/.cache/nixpkgs-review/pr-194714]$ ./results/factplusplus/bin/FaCT++
FaCT++.Kernel: Reasoner for the SROIQ(D) Description Logic, 64-bit
Copyright (C) Dmitry Tsarkov, 2002-2016. Version 1.6.5 (31 December 2016)

Usage:	FaCT++ <Conf file>  or
	FaCT++ -get-default-options


[nix-shell:~/.cache/nixpkgs-review/pr-194714]$ ./results/factplusplus/bin/FaCT++ -get-default-options
FaCT++.Kernel: Reasoner for the SROIQ(D) Description Logic, 64-bit
Copyright (C) Dmitry Tsarkov, 2002-2016. Version 1.6.5 (31 December 2016)
[LeveLogger]

;--- Logging file name
 file = reasoning.log
;--- Logging level (the less level you give, the less information will be logged)
 allowedLevel = 0


[Tuning]

;---
;--- Option 'IAOEFLG': text ---
;---
;* Option 'IAOEFLG' define the priorities of different operations in TODO list. Possible values are 7-digit strings with ony possible digit are 0-6. The digits on the places 1, 2, ..., 7 are for priority of Id, And, Or, Exists, Forall, LE and GE operations respectively. The smaller number means the higher priority. All other constructions (TOP, BOTTOM, etc) has priority 0.
;* Default value: '1263005'

; IAOEFLG = 1263005

;---
;--- Option 'absorptionFlags': text ---
;---
;* Option 'absorptionFlags' sets up absorption process for general axioms. It text field of arbitrary length; every symbol means the absorption action: (B)ottom Absorption), (T)op absorption, (E)quivalent concepts replacement, (C)oncept absorption, (N)egated concept absorption, (F)orall expression replacement, Simple (f)orall expression replacement, (R)ole absorption, (S)plit
;* Default value: 'BTEfCFSR'

; absorptionFlags = BTEfCFSR

;---
;--- Option 'allowUndefinedNames': boolean ---
;---
;* Option 'allowUndefinedNames' describes the policy of undefined names.
;* Default value: 'true'

; allowUndefinedNames = 1

;---
;--- Option 'alwaysPreferEquals': boolean ---
;---
;* Option 'alwaysPreferEquals' allows user to enforce usage of C=D definition instead of C[=D during absorption, even if implication appeares earlier in stream of axioms.
;* Default value: 'true'

; alwaysPreferEquals = 1

;---
;--- Option 'checkAD': boolean ---
;---
;* Option 'checkAD' forces FaCT++ to create the AD and exit instead of performing classification
;* Default value: 'false'

; checkAD = 0

;---
;--- Option 'dumpOntology': boolean ---
;---
;* Option 'dumpOntology' dumps the ontology loaded into the reasoner in a LISP-like format
;* Default value: 'false'

; dumpOntology = 0

;---
;--- Option 'dumpQuery': boolean ---
;---
;* Option 'dumpQuery' dumps sub-TBox relevant to given satisfiability/subsumption query.
;* Default value: 'false'

; dumpQuery = 0

;---
;--- Option 'orSortSat': text ---
;---
;* Option 'orSortSat' define the sorting order of OR vertices in the DAG used in satisfiability tests (used mostly in caching). Option has form of string 'Mop', see orSortSub for details.
;* Default value: '0'

; orSortSat = 0

;---
;--- Option 'orSortSub': text ---
;---
;* Option 'orSortSub' define the sorting order of OR vertices in the DAG used in subsumption tests. Option has form of string 'Mop', where 'M' is a sort field (could be 'D' for depth, 'S' for size, 'F' for frequency, and '0' for no sorting), 'o' is a order field (could be 'a' for ascending and 'd' for descending mode), and 'p' is a preference field (could be 'p' for preferencing non-generating rules and 'n' for not doing so).
;* Default value: '0'

; orSortSub = 0

;---
;--- Option 'skipBeforeBlock': integer ---
;---
;* Internal use only. Option 'skipBeforeBlock' allow user to skip given number of nodes before make a block.
;* Default value: '0'

; skipBeforeBlock = 0

;---
;--- Option 'testTimeout': integer ---
;---
;* Option 'testTimeout' sets timeout for a single reasoning test in milliseconds.
;* Default value: '0'

; testTimeout = 0

;---
;--- Option 'useAnywhereBlocking': boolean ---
;---
;* Option 'useAnywhereBlocking' allow user to choose between Anywhere and Ancestor blocking.
;* Default value: 'true'

; useAnywhereBlocking = 1

;---
;--- Option 'useBackjumping': boolean ---
;---
;* Option 'useBackjumping' switch backjumping on and off. The usage of backjumping usually leads to much faster reasoning.
;* Default value: 'true'

; useBackjumping = 1

;---
;--- Option 'useCompletelyDefined': boolean ---
;---
;* Option 'useCompletelyDefined' leads to simpler Taxonomy creation if TBox contains no non-primitive concepts. Unfortunately, it is quite rare case.
;* Default value: 'true'

; useCompletelyDefined = 1

;---
;--- Option 'useIncrementalReasoning': boolean ---
;---
;* Option 'useIncrementalReasoning' (development) allows one to reason efficiently about small changes in the ontology.
;* Default value: 'false'

; useIncrementalReasoning = 0

;---
;--- Option 'useLazyBlocking': boolean ---
;---
;* Option 'useLazyBlocking' makes checking of blocking status as small as possible. This greatly increase speed of reasoning.
;* Default value: 'true'

; useLazyBlocking = 1

;---
;--- Option 'useSemanticBranching': boolean ---
;---
;* Option 'useSemanticBranching' switch semantic branching on and off. The usage of semantic branching usually leads to faster reasoning, but sometime could give small overhead.
;* Default value: 'true'

; useSemanticBranching = 1

;---
;--- Option 'useSpecialDomains': boolean ---
;---
;* Option 'useSpecialDomains' (development) controls the special processing of R&D for non-simple roles. Should always be set to true.
;* Default value: 'true'

; useSpecialDomains = 1

```

File: pkgs/applications/science/logic/factplusplus/default.nix [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    sed -i 's/OS = MACOSX/OS = LINUX/g' Makefile.include
+    printf '%s\n%s\n' '#include <iostream>' "$(cat Kernel/AtomicDecomposer.cpp)" > Kernel/AtomicDecomposer.cpp
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin
+    mkdir -p $out/lib
+    install -m 755 FaCT++.{C,JNI,KE,Kernel}/obj/*.{so,o} $out/lib/
+    install -m 755 FaCT++/obj/FaCT++ $out/bin/FaCT++
+  '';
```
Comment (h7x4) [inline pkgs/applications/science/logic/factplusplus/default.nix:28]:
```suggestion
  configurePhase = ''
    runHook preConfigure
    
    sed -i 's/OS = MACOSX/OS = LINUX/g' Makefile.include
    printf '%s\n%s\n' '#include <iostream>' "$(cat Kernel/AtomicDecomposer.cpp)" > Kernel/AtomicDecomposer.cpp
    
    runHook postConfigure
  '';

  installPhase = ''
    runHook preInstall
    
    install -Dm755 FaCT++.{C,JNI,KE,Kernel}/obj/*.{so,o} -t $out/lib/
    install -Dm755 FaCT++/obj/FaCT++ -t $out/bin
    
    runHook postInstall
  '';
```

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -3909,6 +3909,8 @@ with pkgs;
 
   facter = callPackage ../tools/system/facter { };
 
+  factplusplus = callPackage ../applications/science/logic/factplusplus { };
```
Comment (h7x4) [inline pkgs/top-level/all-packages.nix:3912]: Would you mind moving the package to [`by-name`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md)?

**Review by h7x4 — COMMENTED** [review]
Ping me if it stales again :)

File: pkgs/by-name/fa/factplusplus/package.nix [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    runHook postInstall
+  '';
+
+   meta = with lib; {
+    description = "Tableaux-based reasoner for expressive Description Logics (DL)";
+    homepage    = "http://owl.cs.manchester.ac.uk/tools/fact/";
+    maintainers = [ maintainers.mgttlinger ];
+    license     = licenses.gpl2Plus;
+    platforms   = platforms.linux;
```
[the remark this answers is not in the document]
Comment (h7x4) [reply pkgs/by-name/fa/factplusplus/package.nix]: It's okay that you're not able to test the other platforms. But AFAIK, `meta.platforms` represents which platforms the (upstream) software supports or what we think it supports if they don't specify anything. Setting `broken = stdenv.hostPlatform.isDarwin` (or alternatively `broken = !stdenv.hostPlatform.isLinux` or `meta.badPlatforms`) would be the way to specify that we currently don't support this platform in nixpkgs (yet).

**Comment by h7x4** [comment]
Also, when you're done, would you mind squashing the commits down into a single commit `factplusplus: init at 1.6.5`?

**Comment by h7x4** [comment]
Thank you!

**Pull request merged**